### PR TITLE
Logger Updates

### PR DESCRIPTION
**Pull request by DannyPeck**
- Added logger convenience methods for logging at a particular severity level
- Implemented the << operator for Sinks, now any subclass of Sink can use the << operator by default and only needs to write the output method (which is required).
- Fixed some style issues due to other discussions

### CODE REVIEW

File: C++/Logger/Logger.cpp [hunk 1]
```diff
@@ -8,14 +7,40 @@ Logger::Logger (Sink & sink)
 : sink_ (sink)
 {}
```
Comment (technetos) [inline C++/Logger/Logger.cpp]: {

}

File: C++/Logger/Logger.cpp [hunk 2]
```diff
@@ -8,14 +7,40 @@ Logger::Logger (Sink & sink)
...
 
-void Logger::log (const Severity::Level & level, const std::string & message)
+void Logger::log (const Severity::Level & level, const std::string & message) const
 {
   if (level >= this->threshold_)
```
Comment (technetos) [inline C++/Logger/Logger.cpp]: `threshhold_` or `_threshold`

File: C++/Logger/SyncFileSink.h [hunk 3]
```diff
@@ -1,25 +0,0 @@
-#ifndef SYNC_FILE_SINK_H
-#define SYNC_FILE_SINK_H
-
-#include "Sink.h"
-#include <mutex>
```
Comment (technetos) [inline C++/Logger/SyncFileSink.h:5]: ...
#include "whatever.h"
#include "Sink.h"

#include \<mutex\>
#include \<system_whatever\>
#Include \<.....

Comment (technetos) [reply C++/Logger/SyncFileSink.h:5]: Its good to separate ours vs theres.

Comment (technetos) [reply C++/Logger/SyncFileSink.h:5]: Also we need to have any of our includes before the system includes

File: C++/Logger/SyncFileSink.h [hunk 4]
```diff
@@ -1,25 +0,0 @@
...
-class SyncFileSink
-{
-public:
-  SyncFileSink (const std::string & filename);
-
-  // Sink
-  void output (const std::string & message);
-
-private:
-  std::string filename_;
```
Comment (technetos) [inline C++/Logger/SyncFileSink.h:19]: see above

File: C++/Logger/SyncFileSink.h [hunk 5]
```diff
@@ -1,25 +0,0 @@
...
-  std::mutex mut_;
```
Comment (technetos) [inline C++/Logger/SyncFileSink.h:20]: ^

File: C++/Logger/sinks/sync/StdOutSink.h [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#ifndef _SYNC_STD_OUT_SINK_H_
+#define _SYNC_STD_OUT_SINK_H_
+
+#include "../Sink.h"
+#include <string>
```
Comment (technetos) [inline C++/Logger/sinks/sync/StdOutSink.h]: See above

File: C++/Logger/sinks/Sink.cpp [hunk 7]
```diff
@@ -0,0 +1,12 @@
+#include "Sink.h"
+
+namespace Toolbox::Log::Sinks
+{
+
+Sink & Sink::operator << (const std::string & message)
+{
+  this->output (message);
+  return *this;
```
Comment (technetos) [inline C++/Logger/sinks/Sink.cpp:9]: return * this

File: C++/Logger/Logger.cpp (hunk 1 above)
Comment (DannyPeck) [reply C++/Logger/Logger.cpp]: is that this:
```
{

}
```
or
```
{
}
```

File: C++/Logger/Logger.cpp (hunk 2 above)
Comment (DannyPeck) [reply C++/Logger/Logger.cpp]: hmm... this->_threshold... I think I actually like it in front tbh..

File: C++/Logger/SyncFileSink.h (hunk 3 above)
Comment (DannyPeck) [reply C++/Logger/SyncFileSink.h:5]: Agreed, I will update it.

File: C++/Logger/sinks/Sink.cpp (hunk 7 above)
Comment (DannyPeck) [reply C++/Logger/sinks/Sink.cpp:9]: I think it is good practice to keep the * and & next to the object when you are doing dereference and address-of.  When it is a pointer declaration, I like the space.

Comment (DannyPeck) [reply C++/Logger/sinks/Sink.cpp:9]: I believe it is good to have the * and & next to the object when it is a dereference and address-of operator.  I agree with the space when it is a pointer declaration.

File: C++/Logger/SyncFileSink.h (hunk 3 above)
Comment (DannyPeck) [reply C++/Logger/SyncFileSink.h:5]: I agree, I will update this.

File: C++/Logger/Logger.cpp (hunk 2 above)
Comment (DannyPeck) [reply C++/Logger/Logger.cpp]: this->_threshold... I honestly like it in front...

File: C++/Logger/Logger.cpp (hunk 1 above)
Comment (technetos) [reply C++/Logger/Logger.cpp]: {

}

Comment (DannyPeck) [reply C++/Logger/Logger.cpp]: are you saying:
```
{

}
```
or
```
{
}
```

File: C++/Logger/Logger.cpp (hunk 2 above)
Comment (technetos) [reply C++/Logger/Logger.cpp]: _threshold

File: C++/Logger/sinks/Sink.cpp (hunk 7 above)
Comment (technetos) [reply C++/Logger/sinks/Sink.cpp:9]: ok.

File: C++/Logger/SyncFileSink.h (hunk 3 above)
Comment (DannyPeck) [reply C++/Logger/SyncFileSink.h:5]: What about something like this?
```
// --------------------------------------------------
#include "StdOutSink.h"
// --------------------------------------------------
#include <iostream>
// --------------------------------------------------
```
We do this at work and I kind of like it, they also do it between each function definition.

File: C++/Logger/Logger.cpp (hunk 1 above)
Comment (technetos) [inline C++/Logger/Logger.cpp]: oh oh. 
`{

}`

Comment (technetos) [inline C++/Logger/Logger.cpp]: `{

}`

File: C++/Logger/SyncFileSink.h (hunk 3 above)
Comment (technetos) [reply C++/Logger/SyncFileSink.h:5]: Its too much.

**Review by technetos — APPROVED** [review]

**Pull request merged**